File: TopCoder/python/arr/SearchInRotatedArray.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if list is None or len(nums) == 0:
            return 0

        low = 0
        high = len(nums) - 1

        while low <= high :
            mid = int(low + (high - low) / 2)

            if nums[mid] == target:
                return mid

            if nums[low] <= nums[mid] :
                if nums[low] <= target <= nums[mid]:
                    high = mid-1
                else:
                    low = mid+1
            else:
                if nums[mid] <= target <= nums[high]:
                    high = mid+1
                else:
                    low = mid-1
        return -1

arr = [4,5,6,7,0,1,2]
sol = Solution()
res = sol.search(arr, 0)
print(res)

File: TopCoder/src/main/java/com/leetcode/array/leetcode/SearchInRotatedArray.java
package com.leetcode.array.leetcode;

//https://leetcode.com/problems/search-in-rotated-sorted-array/
public class SearchInRotatedArray {

    
    public int search(int[] nums, int target) {

        int result = searchRecursive(nums, 0, nums.length - 1, target);
        return result;
    }

    private int searchRecursive(int[] nums, int low, int high, int target) {
        if (low > high) {
            return  -1;
        }
        if (low == high) {
            if (nums[low] == target) {
                return low;
            }
            return -1;
        }


        if (low + 1 == high) {
            if (nums[low] == target) {
                return low;
            } else if (nums[high] == target) {
                return high;
            }
        }

        int mid = (low + high)/2;
        if (nums[mid] == target) {
            return mid;
        }

        if (nums[low] >= nums[mid] && target <= nums[mid] && target <= nums[low]) {
            // pivot is in left
            return searchRecursive(nums, low, mid, target);
        } else if (nums[low] >= nums[mid] && target >= nums[mid] && target >= nums[low]) {
            // pivot is in left
            return searchRecursive(nums, low, mid, target);
        } else if (nums[low] < nums[mid] && target <= nums[mid] && target >= nums[low]) {
            return searchRecursive(nums, low, mid, target);
        } else {
            return searchRecursive(nums, mid + 1, high, target);
        }

    }


    public static void main(String[] args) {
        int nums[] = {5, 1, 3};
        int target = 3;
        int result = new SearchInRotatedArray().search(nums, target);
        System.out.println(result);
    }
}
